File: bank-service/src/main/java/rs/edu/raf/mapper/CreditMapper.java
package rs.edu.raf.mapper;

import lombok.AllArgsConstructor;
import org.springframework.stereotype.Component;
import rs.edu.raf.dto.CreditDTO;
import rs.edu.raf.model.credit.Credit;
import rs.edu.raf.model.credit.CreditRequest;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneId;

@Component
@AllArgsConstructor
public class CreditMapper {
    private CreditRequestMapper creditRequestMapper;

    public Credit creditRequestToCredit(CreditRequest creditRequest) {
        Credit credit = new Credit();

        credit.setCreditRequest(creditRequest);
        credit.setContractDate(System.currentTimeMillis());

        long loanMaturityDate = LocalDateTime
                .ofInstant(Instant.ofEpochMilli(System.currentTimeMillis()), ZoneId.systemDefault())
                .plusMonths(creditRequest.getLoanTerm())
                .atZone(ZoneId.systemDefault())
                .toInstant()
                .toEpochMilli();

        credit.setLoanMaturityDate(loanMaturityDate);
        credit.setRemainingDebt(creditRequest.getLoanAmount());

        long nextInstallmentDate = LocalDateTime
                .ofInstant(Instant.ofEpochMilli(System.currentTimeMillis()), ZoneId.systemDefault())
                .plusMonths(1)
                .atZone(ZoneId.systemDefault())
                .toInstant()
                .toEpochMilli();

        credit.setNextInstallmentDate(nextInstallmentDate);

        return credit;
    }

    public CreditDTO creditToCreditDTO(Credit credit) {
        return new CreditDTO(creditRequestMapper.creditRequestToCreditRequestDto(credit.getCreditRequest()),
                credit.getContractDate(), credit.getLoanMaturityDate(), credit.getInstallmentAmount(),
                credit.getRemainingDebt(), credit.getNextInstallmentDate());
    }
}

File: bank-service-kotlin/src/main/kotlin/org/example/bankservicekotlin/mapper/CreditMapper.kt
package org.example.bankservicekotlin.mapper

import org.example.bankservicekotlin.model.credit.Credit
import org.example.bankservicekotlin.model.credit.CreditRequest
import org.example.bankservicekotlin.dto.CreditDTO
import org.springframework.stereotype.Component
import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId

@Component
class CreditMapper(private val creditRequestMapper: CreditRequestMapper) {
    fun creditRequestToCredit(creditRequest: CreditRequest): Credit {
        val loanMaturityDate = LocalDateTime
            .ofInstant(Instant.ofEpochMilli(System.currentTimeMillis()), ZoneId.systemDefault())
            .plusMonths(creditRequest.loanTerm)
            .atZone(ZoneId.systemDefault())
            .toInstant()
            .toEpochMilli()

        val nextInstallmentDate = LocalDateTime
            .ofInstant(Instant.ofEpochMilli(System.currentTimeMillis()), ZoneId.systemDefault())
            .plusMonths(1)
            .atZone(ZoneId.systemDefault())
            .toInstant()
            .toEpochMilli()

        return Credit(
            null,
            creditRequest,
            System.currentTimeMillis(),
            loanMaturityDate,
            creditRequest.loanAmount,
            creditRequest.loanAmount,
            nextInstallmentDate
        )
    }

    fun creditToCreditDTO(credit: Credit): CreditDTO {
        return CreditDTO(
            creditRequestMapper.creditRequestToCreditRequestDto(credit.creditRequest),
            credit.contractDate, credit.loanMaturityDate, credit.installmentAmount,
            credit.remainingDebt, credit.nextInstallmentDate
        )
    }
}